File: src/api/Dtos/Responses/ApiResult.cs
using Example.Api.Enums;

namespace Example.Api.Dtos.Responses;

/// <summary>
/// Represents the result of an API or Service operation.
/// </summary>
/// <value></value>
public record ApiResult
{
    /// <summary>
    /// Indicates whether the operation was successful.
    /// </summary>
    /// <value></value>
    public bool Success { get; init; }

    /// <summary>
    /// The code representing the result of the operation.
    /// </summary>
    /// <value></value>
    public ApiCode Code { get; init; }

    /// <summary>
    /// The message providing additional information about the result.
    /// </summary>
    /// <value></value>
    public string? Message { get; init; } = string.Empty;
}

/// <summary>
/// Represents the result of an API or Service operation with data.
/// </summary>
/// <value></value>
public record ApiResult<T> : ApiResult
{
    /// <summary>
    /// The data returned by the operation.
    /// </summary>
    /// <value></value>
    public T? Data { get; init; }
}

File: src/nodejs/src/dtos/ApiResult.ts
import { ApiCode } from "./ApiCode";

/**
 * Base class for API responses.
 */
export class ApiResult {
  /** Indicates whether the operation was successful. */
  success: boolean;
  /** The API response code. */
  code: ApiCode;
  /** A descriptive message about the operation result. */
  message: string;

  /**
   * Creates a new API result.
   * @param success - Whether the operation was successful.
   * @param code - The API response code.
   * @param message - A descriptive message about the operation result.
   */
  constructor(success: boolean, code: ApiCode, message: string = "") {
    this.success = success;
    this.code = code;
    this.message = message;
  }
}

/**
 * API response containing data of type T.
 * @template T - The type of data contained in the response.
 */
export class ApiDataResult<T> extends ApiResult {
  /** The data returned by the API operation. */
  data?: T;

  /**
   * Creates a new API data result.
   * @param success - Whether the operation was successful.
   * @param code - The API response code.
   * @param message - A descriptive message about the operation result.
   * @param data - The data returned by the operation.
   */
  constructor(success: boolean, code: ApiCode, message: string = "", data?: T) {
    super(success, code, message);
    this.data = data;
  }
}
